### ISSUE TITLE
pre-built function - to convert timex expression into a UTC ISO format

### ISSUE THREAD

**Issue by VinaiMeyyappan** [microsoft/botbuilder-dotnet#4425]
## Version
What package version of the SDK are you using.
v4

## Describe the bug
Give a clear and concise description of what the bug is.
- Collect date and time in separate steps of a descriptive adaptive dialog 
- Combine them into UTC ISO format before calling a REST API with HTTPRequest action. 
> If a user enters today and 2:30 PM for date and time prompts respectively. 
- the following expression will return - 2020-08-10T14:30:00.000Z
"${formatDateTime(concat($ReadingDate.timex[0], $ReadingTime.timex[0]))}"
> if a user enters today and 2 PM, the bot throws the following error 
- Sorry, it looks like something went wrong.
- XXXXX.en-us.lg:Could not parse 2020-08-11T14 temp Error occurred when evaluating '-${formatDateTime(concat($ReadingDate.timex0, $ReadingTime.timex0))}. 

.

## Expected behavior
Give a clear and concise description of what you expected to happen.

## Additional context

[bug]

**Comment by carlosscastro** [comment]
@boydc2014 Could you verify the repro and triage / redirect? Thanks!

**Comment by carlosscastro** [comment]
@boydc2014 kind reminder :)

**Comment by cosmicshuai** [comment]
I tested several cases of the timestamp format.
for formatDateTime function, `2020-08-10T14` is not a valid input. However, `2020-08-10T14:00` is valid.
My opinion is we should have a solution to make sure  `$ReadingTime.timex[0]` can be converted to `THH:mm`  or `THH:mm:ss` formats correctly. Could you provide more information about the type of `$ReadingTime`. If it is a Timex-Property object, we could add some prebulit functions to do the conversions between Timex-Property and date or time.

**Comment by peterinnesmsft** [comment]
@boydc2014 is there any update on initial triage for this issue?

**Comment by boydc2014** [comment]
@peterinnesmsft Shuai and I was looking into this last week.  Shuai have his initial reply above.


So, @VinaiMeyyappan, our investigation showed that "2020-08-11T14" is not a valid timestamp format using the standardized C# library. Can you share more details about how you get this timestamp, is it using timex library or LUIS?  if it's a common scenario, like @cosmicshuai proposed above, we will add some pre-built functions to help on the convert.

**Comment by Danieladu** [comment]
Timex is a general type for date/time/daterange/timerange/datetimerange, if you just want to get the "datetime" type,  suggest you use the "value" property in LUIS resolution. That might work.

**Comment by peterinnesmsft** [comment]
@VinaiMeyyappan is there any additional information or asks on the team to help resolve this issue that you are looking for?

**Comment by Danieladu** [comment]
I can provide some suggestions for the datetime input usage:
Here is the dialog file, as as reference.
```

    "$schema": "../../../tests.schema",
    "$kind": "Microsoft.Test.Script",
    "dialog": {
        "$kind": "Microsoft.AdaptiveDialog",
        "id": "planningTest",
        "triggers": [
            {
                "$kind": "Microsoft.OnUnknownIntent",
              "actions": [
                {
                  "$kind": "Microsoft.DateTimeInput",
                  "property": "$ReadingDate",
                  "prompt": "Please enter a date.",
                  "outputFormat": "=this.value[0].Value"
                },
                {
                  "$kind": "Microsoft.SendActivity",
                  "activity": "date: ${$ReadingDate}"
                },
                {
                  "$kind": "Microsoft.DateTimeInput",
                  "property": "$ReadingTime",
                  "prompt": "Please enter a time.",
                  "outputFormat": "=this.value[0].Value"
                },
                {
                  "$kind": "Microsoft.SendActivity",
                  "activity": "time: ${$ReadingTime}"
                },
                {
                  "$kind": "Microsoft.SendActivity",
                  "activity": "${formatDateTime(concat($ReadingDate, ' ', $ReadingTime), 'MM-dd-yy')}"
                }
              ]
            }
        ],
        "defaultResultProperty": "dialog.result"
    },
    "script": [
        {
            "$kind": "Microsoft.Test.UserSays",
            "text": "hi"
        },
        {
            "$kind": "Microsoft.Test.AssertReply",
            "text": "Please enter a date."
        },
        {
            "$kind": "Microsoft.Test.UserSays",
            "text": "Aug 19 2020"
        },
        {
            "$kind": "Microsoft.Test.AssertReply",
            "text": "date: 2020-08-19"
        },
        {
            "$kind": "Microsoft.Test.AssertReply",
            "text": "Please enter a time."
        },
      {
        "$kind": "Microsoft.Test.UserSays",
        "text": "2 PM"
      },
      {
        "$kind": "Microsoft.Test.AssertReply",
        "text": "time: 14:00:00"
      },
      {
        "$kind": "Microsoft.Test.AssertReply",
        "text": "08-19-20"
      }
    ]
}
```

Thanks.

**Comment by VinaiMeyyappan** [comment]
use this schema file

`{
    "$schema": "http://json-schema.org/draft-07/schema",
    "type": "object",
    "properties": {
        "ReadingDate": {
            "type": "string",
            "format": "date"
        },
        "ReadingTime": {
            "type": "string",
            "format": "time"
        }
    },
     "required": [
        "ReadingDate",
        "ReadingTime"
    ]
}
`
follow the [instructions ](https://github.com/microsoft/BotBuilder-Samples/blob/master/experimental/generation/docs/get-started.md)to generate, build and run the bot

Look for this code block in the generated dialog file
`{
      "$kind": "Microsoft.OnEndOfActions",
      "condition": "empty(where(dialog.requiredProperties, property, !dialog[property]))",
      "priority": 1000,
      "actions": [
`
Add the lines below to actions
` 
        {
          "$kind": "Microsoft.SendActivity",
          "activity": "${formatDateTime(concat($ReadingDate.timex[0], $ReadingTime.timex[0]))}"
        },
`
Run the bot

> when prompted for date 
> enter today
> when prompted for time
> enter  2 PM

Sorry, it looks like something went wrong.
XXXXX.en-us.lg:Could not parse 2020-08-19T14 temp Error occurred when evaluating '-${formatDateTime(concat($ReadingDate.timex0, $ReadingTime.timex0))}.

**Comment by boydc2014** [comment]
> use this schema file
> 
> `{ "$schema": "http://json-schema.org/draft-07/schema", "type": "object", "properties": { "ReadingDate": { "type": "string", "format": "date" }, "ReadingTime": { "type": "string", "format": "time" } }, "required": [ "ReadingDate", "ReadingTime" ] } `
> follow the [instructions ](https://github.com/microsoft/BotBuilder-Samples/blob/master/experimental/generation/docs/get-started.md)to generate, build and run the bot
> 
> Look for this code block in the generated dialog file
> `{ "$kind": "Microsoft.OnEndOfActions", "condition": "empty(where(dialog.requiredProperties, property, !dialog[property]))", "priority": 1000, "actions": [ `
> Add the lines below to actions
> ` { "$kind": "Microsoft.SendActivity", "activity": "${formatDateTime(concat($ReadingDate.timex[0], $ReadingTime.timex[0]))}" },`
> Run the bot
> 
> > when prompted for date
> > enter today
> > when prompted for time
> > enter  2 PM
> 
> Sorry, it looks like something went wrong.
> XXXXX.en-us.lg:Could not parse 2020-08-19T14 temp Error occurred when evaluating '-${formatDateTime(concat($ReadingDate.timex0, $ReadingTime.timex0))}.

Thanks @VinaiMeyyappan  for the infomation, we will try to setup the environment.  

The root cause as above is because this timex value (2020-08-19T14) is not a standardized date time format, thus can't be formatted in current implemented.  

But per our previous investigation, i think both LuisRecognizer and DateTimeInput already done this formatting, they will has a "value" property which already holds the resolved value of datetime, i will check more details on this.   Are you developing those generated scripts or you just using that? Can you try use the resolved, instead of the direct timex0 property?

**Comment by boydc2014** [comment]
Hi, @VinaiMeyyappan, as @Danieladu stated above, the direct cause of this issue is FormatDatetime in adapativeExpression was not designed format a timex string.  

To get a timex value resolved, we can either rely on the LUIS result, which unfortunately we dropped in the processing pipeline by default, or we will create new pre-built functions to specific for formating timex value. 

I will leave the decision of whether we want to expose LUIS resolution to user layer to @tomlm and @chrimc62, i know there is a reason we chose to drop than. For now we will go with the latter approach to add functions for that first. 

Will get back to you once the functions are been added.
